RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+#include <iostream>
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+#include "CondFormats/SiPixelTransient/interface/SiPixelTemplate.h"
+#include "DataFormats/DetId/interface/DetId.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+#include "Geometry/TrackerGeometryBuilder/interface/RectangularPixelTopology.h"
+#include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEFast.h"
+
+// Services
+// this is needed to get errors from templates
+
+namespace {
+  constexpr float micronsToCm = 1.0e-4;
+}
+
+//-----------------------------------------------------------------------------
+//!  The constructor.
+//-----------------------------------------------------------------------------
+PixelCPEFast::PixelCPEFast(edm::ParameterSet const& conf,
+                           const MagneticField* mag,
+                           const TrackerGeometry& geom,
+                           const TrackerTopology& ttopo,
+                           const SiPixelLorentzAngle* lorentzAngle,
+                           const SiPixelGenErrorDBObject* genErrorDBObject,
+                           const SiPixelLorentzAngle* lorentzAngleWidth)
+    : PixelCPEBase(conf, mag, geom, ttopo, lorentzAngle, genErrorDBObject, nullptr, lorentzAngleWidth, 0) {
+  EdgeClusterErrorX_ = conf.getParameter<double>("EdgeClusterErrorX");
+  EdgeClusterErrorY_ = conf.getParameter<double>("EdgeClusterErrorY");
+
+  UseErrorsFromTemplates_ = conf.getParameter<bool>("UseErrorsFromTemplates");
+  TruncatePixelCharge_ = conf.getParameter<bool>("TruncatePixelCharge");
+
+  // Use errors from templates or from GenError
+  if (UseErrorsFromTemplates_) {
+    if (!SiPixelGenError::pushfile(*genErrorDBObject_, thePixelGenError_))
+      throw cms::Exception("InvalidCalibrationLoaded")
+          << "ERROR: GenErrors not filled correctly. Check the sqlite file. Using SiPixelTemplateDBObject version "
+          << (*genErrorDBObject_).version();
+  }
+
+  // Rechit errors in case other, more correct, errors are not vailable
+  // This are constants. Maybe there is a more efficienct way to store them.
+  xerr_barrel_l1_ = {0.00115, 0.00120, 0.00088};
+  xerr_barrel_l1_def_ = 0.01030;
+  yerr_barrel_l1_ = {0.00375, 0.00230, 0.00250, 0.00250, 0.00230, 0.00230, 0.00210, 0.00210, 0.00240};
+  yerr_barrel_l1_def_ = 0.00210;
+  xerr_barrel_ln_ = {0.00115, 0.00120, 0.00088};
+  xerr_barrel_ln_def_ = 0.01030;
+  yerr_barrel_ln_ = {0.00375, 0.00230, 0.00250, 0.00250, 0.00230, 0.00230, 0.00210, 0.00210, 0.00240};
+  yerr_barrel_ln_def_ = 0.00210;
+  xerr_endcap_ = {0.0020, 0.0020};
+  xerr_endcap_def_ = 0.0020;
+  yerr_endcap_ = {0.00210};
+  yerr_endcap_def_ = 0.00075;
```

Comment:
these should be (eventually) factored out in a common place to be reused here and in `PixelCPEGeneric`.

What about phase-2 support?
If that's going to be handled by another class, I think that this one should be renamed to `PixelPhase1CPEFast` or similar

Reply:
Added to the post-merge open issues at https://github.com/cms-sw/cmssw/issues/32483 .